src/pd/leader.rs

```diff
@@ -1,51 +1,222 @@
+        let tso_rx = client.wl().reactor.tso_rx.take().unwrap();
+        handle.spawn(
+            tx.sink_map_err(Error::Grpc)
+                .send_all(tso_rx.then(|r| match r {
+                    Ok(r) => Ok((r, WriteFlags::default())),
+                    Err(()) => Err(Error::Other(box_err!("failed to recv tso requests"))),
+                })).then(|r| match r {
+                    Ok((mut sender, _)) => {
+                        sender.get_mut().cancel();
+                        Ok(())
+                    }
+                    Err(e) => {
+                        error!("failed to send tso requests: {:?}", e);
+                        Err(())
+                    }
+                }),
+        );
+        handle.spawn(
+            rx.for_each(move |resp| {
+                let mut client = client.wl();
+                let reactor = &mut client.reactor;
+                let tso_pending = reactor.tso_pending.take().unwrap();
+                reactor.schedule(PdTask::TsoResponse(tso_pending, resp));
+                if !reactor.tso_batch.is_empty() {
+                    /* schedule another tso_batch of request */
+                    reactor.schedule(PdTask::TsoRequest);
+                }
+                Ok(())
+            }).map_err(|e| panic!("unexpected error: {:?}", e)),
+        );
+    }
+
+    fn tso_request(client: &Arc<RwLock<LeaderClient>>) {
+        let mut client = client.wl();
+        let cluster_id = client.cluster_id;
+        let reactor = &mut client.reactor;
+        let mut tso_batch = reactor.tso_buffer.take().unwrap();
+        tso_batch.extend(reactor.tso_batch.drain(..));
+        let mut request = request!(cluster_id, TsoRequest);
+        request.set_count(tso_batch.len() as u32);
+        reactor.tso_pending = Some(tso_batch);
+        reactor.tso_tx.unbounded_send(request).unwrap();
+    }
+
+    fn tso_response(
+        client: &Arc<RwLock<LeaderClient>>,
+        mut requests: Vec<TsoChannel>,
+        response: &TsoResponse,
+    ) {
+        let timestamp = response.get_timestamp();
+        for (offset, request) in requests.drain(..).enumerate() {
+            request
+                .send(PdTimestamp {
+                    physical: timestamp.physical,
+                    logical: timestamp.logical + offset as i64,
+                }).unwrap();
+        }
+        client.wl().reactor.tso_buffer = Some(requests);
+    }
+
+    fn dispatch(client: &Arc<RwLock<LeaderClient>>, task: PdTask, handle: &OtherHandle) {
+        match task {
+            PdTask::TsoRequest => Self::tso_request(client),
+            PdTask::TsoResponse(requests, response) => Self::tso_response(client, requests, &response),
+            PdTask::TsoInit => Self::init(client, handle),
+            PdTask::Request(task) => handle.spawn(task),
+        }
+    }
+
+    fn get_ts(&mut self) -> PdFuture<PdTimestamp> {
+        let timer = Instant::now();
+        let (tx, rx) = oneshot::channel::<PdTimestamp>();
```

Comment:
You asked me to comment on this. :)

It seems creating this channel is non-trivially expensive:

```rust
#![feature(test)]

extern crate test;
extern crate futures;

use futures::{Future, sync::oneshot};
use test::Bencher;
use std::time::Instant;

#[bench]
fn create_channel(b: &mut Bencher) {
    b.iter(|| {
        oneshot::channel::<(i64, i64)>()
    });
}

#[bench]
fn create_channel_and_send_recieve(b: &mut Bencher) {
    b.iter(|| {
        let (tx, rx) = oneshot::channel::<(i64, i64)>();
        tx.send((0,0));
        rx.wait()
    });
}

#[bench]
fn instant(b: &mut Bencher) {
    b.iter(|| {
        Instant::now()
    });
}
```

```rust
running 3 tests
test create_channel                  ... bench:          96 ns/iter (+/- 4)
test create_channel_and_send_recieve ... bench:         141 ns/iter (+/- 9)
test instant                         ... bench:          17 ns/iter (+/- 2)
```

I was curious and looked at `std::sync::atomic::AtomicIsize` since `PdTimestamp` is essentially a `(i64, i64)`, it is predictably much faster:

```rust
#[bench]
fn atomics(b: &mut Bencher) {
    b.iter(|| {
        let values = Arc::new((AtomicIsize::new(0), AtomicIsize::new(0)));
        values.0.store(5, Ordering::Relaxed);
        values.1.store(6, Ordering::Relaxed);
        (values.0.load(Ordering::Relaxed), values.1.load(Ordering::Relaxed));
    });
}
```

```
test atomics                         ... bench:          25 ns/iter (+/- 3)
```

However I think we will pay more in other places because synchronization becomes harder. A similar problem happens with `Once`. Since you have a better context to the code though, maybe you this gives you an idea I don't see.  :)

Reply:
It's strange that I can't find my earlier comments. 

In order to support requests batching and keep the order of responses, I used a dedicated thread to run event loop like what Go pd client does. I haven't figured out other way to implement tso without dispatching thread. Since I already have a thread running I thought it might be good idea to dispatch everything with this thread so there is only one thread interacting with gRPC channel. 

This tso streaming thing really bothered me a lot, would like to hear comments from you and everybody else. Let's see if there could be better way.

Reply:
@Hoverbear Do you have any comment to this?

Reply:
If current implementation is acceptable, I would like to merge this request so I can start a new merge request for new code.